File: rewrite-java/src/test/kotlin/org/openrewrite/java/tree/CursorTest.kt
/*
 * Copyright 2020 the original author or authors.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.java.tree

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.openrewrite.Tree
import org.openrewrite.java.JavaParser
import org.openrewrite.java.RetrieveCursor
import org.openrewrite.java.JavaSourceVisitor

interface CursorTest {
    @Test
    fun inSameNameScope(jp: JavaParser) {
        val a = jp.parse("""
            public class A extends B {
                int n;
                
                public void foo(int n1) {
                    for(int n2 = 0;;) {
                    }
                }
                
                static class B {
                    public void foo(int n) {
                    }
                }
                
                interface C {
                    void foo(int n);
                }
                
                enum D {
                    D1, D2;
                    void foo(int n) {}
                }
                
                class E {
                    void foo(int n) {}
                }
            }
        """.trimIndent())

        fun Tree.cursor() = RetrieveCursor(this).visit(a)

        val fieldScope = a.classes[0].fields[0].cursor()!!
        val methodParamScope = a.classes[0].methods[0].params.params[0].cursor()!!
        val forInitScope = a.classes[0].methods[0].body!!.statements.filterIsInstance<J.ForLoop>()[0].control.init.cursor()!!

        assertThat(object : JavaSourceVisitor<Int>() {
            init {
                setCursoringOn()
            }

            override fun defaultTo(t: Tree?): Int = 0

            override fun visitCompilationUnit(cu: J.CompilationUnit?): Int {
                assertTrue(isInSameNameScope(fieldScope, methodParamScope))
                assertFalse(isInSameNameScope(methodParamScope, fieldScope))
                assertTrue(isInSameNameScope(fieldScope, forInitScope))

                val innerClasses = a.classes[0].body.statements.filterIsInstance<J.ClassDecl>()
                assertEquals(4, innerClasses.size)

                innerClasses.forEachIndexed { n, innerClass ->
                    val innerStaticClassMethodParam = innerClass.methods[0].params.params[0].cursor()!!
                    assertEquals(n >= 3, isInSameNameScope(fieldScope, innerStaticClassMethodParam))
                }

                return 1
            }
        }.visit(a)).isEqualTo(1)
    }
}

File: rewrite-go/test/cursor_test.go
/*
 * Copyright 2025 the original author or authors.
 *
 * Licensed under the Moderne Source Available License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://docs.moderne.io/licensing/moderne-source-available-license
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package test

import (
	"testing"

	"github.com/stretchr/testify/require"

	"github.com/openrewrite/rewrite/rewrite-go/pkg/parser"
	"github.com/openrewrite/rewrite/rewrite-go/pkg/recipe"
	"github.com/openrewrite/rewrite/rewrite-go/pkg/tree/golang"
	"github.com/openrewrite/rewrite/rewrite-go/pkg/tree/java"
	"github.com/openrewrite/rewrite/rewrite-go/pkg/visitor"
)

func TestCursorBuildChain(t *testing.T) {
	cu, err := parser.NewGoParser().Parse("a.go", "package main\n")
	require.NoError(t, err)
	chain := visitor.BuildChain([]java.Tree{cu})
	require.Falsef(t, chain == nil || chain.Value() != cu || chain.Parent() != nil, "expected single-element chain rooted at cu; got parent=%v value", chain.Parent())

	chain2 := visitor.BuildChain(nil)
	require.Nil(t, chain2, "expected nil chain for empty input")
}

// TestVisitorCursorState confirms that GoVisitor exposes its cursor as
// state via Cursor() / SetCursor(), matching the JavaVisitor pattern.
// The RPC layer seeds an initial cursor before traversal; recipes read
// it from inside any Visit* override.
func TestVisitorCursorState(t *testing.T) {
	cu, err := parser.NewGoParser().Parse("a.go", "package main\nfunc f(){}\n")
	require.NoError(t, err)

	v := &cursorObservingVisitor{}
	visitor.Init(v)

	outer := visitor.BuildChain([]java.Tree{cu})
	v.SetCursor(outer)
	if v.Cursor() != outer {
		t.Fatalf("Cursor() should return what SetCursor seeded")
	}

	v.Visit(cu, recipe.NewExecutionContext())
	require.True(t, v.observedCU, "VisitCompilationUnit was never invoked")
	require.False(t, v.cuCursor == nil || v.cuCursor.Value() != cu, "expected v.Cursor().Value() == cu inside VisitCompilationUnit")
}

type cursorObservingVisitor struct {
	visitor.GoVisitor
	observedCU bool
	cuCursor   *visitor.Cursor
}

func (v *cursorObservingVisitor) VisitCompilationUnit(cu *golang.CompilationUnit, p any) java.J {
	v.observedCU = true
	v.cuCursor = v.Cursor()
	return cu
}
